File: AIM-OS/packages/apoe/integration_tags.py
"""Integration Tag Helpers for APOE."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional


@dataclass
class IntegrationSystem:
    name: str
    priority: Optional[str] = None


@dataclass
class IntegrationTagContext:
    system: Optional[IntegrationSystem] = None
    integration_type: Optional[str] = None
    connection: Optional[str] = None
    modality: Optional[str] = None
    action: Optional[str] = None
    mode: Optional[str] = None
    agent: Optional[str] = None
    extras: List[str] = field(default_factory=list)


def build_integration_tags(context: IntegrationTagContext) -> List[str]:
    """Convert context into standardized tag list."""
    tags: List[str] = []

    if context.system:
        priority = context.system.priority or "p0"
        tags.append(f"system:{context.system.name}:{priority}")

    if context.integration_type:
        tags.append(f"integration_type:{context.integration_type}")

    if context.connection:
        tags.append(f"connection:{context.connection}")

    if context.modality:
        tags.append(f"modality:{context.modality}")

    if context.action:
        tags.append(f"action:{context.action}")

    if context.mode:
        tags.append(f"mode:{context.mode}")

    if context.agent:
        tags.append(f"agent:{context.agent}")

    if context.extras:
        tags.extend([extra for extra in context.extras if extra])

    if "chat_ide" not in tags:
        tags.append("chat_ide")

    # Deduplicate
    return list(dict.fromkeys(tags))


def integration_tags_to_dict(tags: List[str]) -> Dict[str, float]:
    """Weight tags for CMC storage."""
    weights: Dict[str, float] = {}

    for tag in tags:
        weight = 0.9
        if tag.startswith("system:"):
            weight = 1.0
        elif tag.startswith("integration_type:"):
            weight = 1.0
        elif tag.startswith("modality:"):
            weight = 1.0
        elif tag.startswith("connection:"):
            weight = 0.9
        elif tag == "chat_ide":
            weight = 1.0
        elif tag.startswith(("action:", "mode:", "agent:")):
            weight = 0.85
        else:
            weight = 0.8

        weights[tag] = weight

    return weights


def merge_integration_context(
    base: Optional[IntegrationTagContext],
    override: Optional[IntegrationTagContext]
) -> IntegrationTagContext:
    """Merge base context with overrides."""
    merged = IntegrationTagContext()

    merged.system = _merge_system(base.system if base else None, override.system if override else None)
    merged.integration_type = override.integration_type or (base.integration_type if base else None)
    merged.connection = override.connection or (base.connection if base else None)
    merged.modality = override.modality or (base.modality if base else None)
    merged.action = override.action or (base.action if base else None)
    merged.mode = override.mode or (base.mode if base else None)
    merged.agent = override.agent or (base.agent if base else None)

    extras: List[str] = []
    if base and base.extras:
        extras.extend(base.extras)
    if override and override.extras:
        extras.extend(override.extras)
    merged.extras = list(dict.fromkeys([extra for extra in extras if extra]))

    return merged


def _merge_system(
    base: Optional[IntegrationSystem],
    override: Optional[IntegrationSystem]
) -> Optional[IntegrationSystem]:
    if not base and not override:
        return None

    name = override.name if override and override.name else (base.name if base else None)
    priority = override.priority if override and override.priority else (base.priority if base else None)

    if not name:
        return None

    return IntegrationSystem(name=name, priority=priority)

File: AIM-OS/ide_orchestration/prototypes/dac/src/utils/integrationTags.ts
/**
 * Integration Tag Helpers
 * Converts chat/IDE orchestration metadata into standardized tag payloads.
 */

export type IntegrationPriority =
  | 'critical'
  | 'important'
  | 'routine'
  | 'low_stakes'
  | 'p0'
  | 'p1'
  | string

export interface IntegrationSystem {
  name: string
  priority?: IntegrationPriority
}

export interface IntegrationTagContext {
  system?: IntegrationSystem
  integrationType?: string
  connection?: string
  modality?: string
  /**
   * Action identifier (e.g., "code_generation", "research_query").
   */
  action?: string
  /**
   * Active thinking mode (research, execution, synthesis, planning, etc.).
   */
  mode?: string
  /**
   * Active drawer / agent identity (coding, planning, emotional, etc.).
   */
  agent?: string
  /**
   * Additional raw tags (chat_ide, session ids, etc.).
   */
  extras?: string[]
}

const REQUIRED_BASE_TAG = 'chat_ide'

/**
 * Build the canonical list-based tag payload given orchestration context.
 */
export function buildIntegrationTags(context: IntegrationTagContext): string[] {
  const tags: string[] = []

  if (context.system?.name) {
    const priority = context.system.priority || 'p0'
    tags.push(`system:${context.system.name}:${priority}`)
  }

  if (context.integrationType) {
    tags.push(`integration_type:${context.integrationType}`)
  }

  if (context.connection) {
    tags.push(`connection:${context.connection}`)
  }

  if (context.modality) {
    tags.push(`modality:${context.modality}`)
  }

  // Context tags (lighter weight but always helpful)
  if (context.action) {
    tags.push(`action:${context.action}`)
  }

  if (context.mode) {
    tags.push(`mode:${context.mode}`)
  }

  if (context.agent) {
    tags.push(`agent:${context.agent}`)
  }

  if (context.extras?.length) {
    tags.push(...context.extras.filter(Boolean))
  }

  if (!tags.includes(REQUIRED_BASE_TAG)) {
    tags.push(REQUIRED_BASE_TAG)
  }

  return Array.from(new Set(tags))
}

/**
 * Convert tag list format into CMC weighted dictionary format.
 * Mirrors Atlas' recommendation to weight connections/context slightly lower
 * than system/integration identifiers for HHNI indexing.
 */
export function integrationTagsToDict(tags: string[]): Record<string, number> {
  const weights: Record<string, number> = {}

  for (const tag of tags) {
    let weight = 0.9

    if (tag.startsWith('system:')) {
      weight = 1.0
    } else if (tag.startsWith('integration_type:')) {
      weight = 1.0
    } else if (tag.startsWith('modality:')) {
      weight = 1.0
    } else if (tag.startsWith('connection:')) {
      weight = 0.9
    } else if (tag === REQUIRED_BASE_TAG) {
      weight = 1.0
    } else if (tag.startsWith('action:') || tag.startsWith('mode:') || tag.startsWith('agent:')) {
      weight = 0.85
    } else {
      weight = 0.8
    }

    weights[tag] = weight
  }

  return weights
}

const normalizeExtras = (extras?: string[]): string[] | undefined => {
  if (!extras || !extras.length) return undefined
  const deduped = Array.from(new Set(extras.filter(Boolean)))
  return deduped.length ? deduped : undefined
}

const mergeSystems = (
  base?: IntegrationSystem,
  override?: IntegrationSystem
): IntegrationSystem | undefined => {
  if (!base && !override) {
    return undefined
  }

  return {
    name: override?.name ?? base?.name ?? 'unknown',
    priority: override?.priority ?? base?.priority
  }
}

export function mergeIntegrationContexts(
  base?: IntegrationTagContext | null,
  override?: IntegrationTagContext | null
): IntegrationTagContext | undefined {
  if (!base && !override) {
    return undefined
  }

  const merged: IntegrationTagContext = {
    system: mergeSystems(base?.system, override?.system),
    integrationType: override?.integrationType ?? base?.integrationType,
    connection: override?.connection ?? base?.connection,
    modality: override?.modality ?? base?.modality,
    action: override?.action ?? base?.action,
    mode: override?.mode ?? base?.mode,
    agent: override?.agent ?? base?.agent,
    extras: normalizeExtras([...(base?.extras || []), ...(override?.extras || [])])
  }

  return merged
}

let activeIntegrationContext: IntegrationTagContext | null = null

export function setActiveIntegrationContext(context: IntegrationTagContext | null): void {
  activeIntegrationContext = context
    ? {
        ...context,
        system: context.system ? { ...context.system } : undefined,
        extras: normalizeExtras(context.extras)
      }
    : null
}

export function getActiveIntegrationContext(): IntegrationTagContext | null {
  return activeIntegrationContext
}

/**
 * Merge globally active context, request-specific context, and action overrides.
 */
export function resolveIntegrationContext(
  requestContext?: IntegrationTagContext,
  overrides?: IntegrationTagContext
): IntegrationTagContext | undefined {
  const base = mergeIntegrationContexts(getActiveIntegrationContext(), requestContext)
  return mergeIntegrationContexts(base, overrides)
}
